model/command_autocomplete.go

```diff
@@ -0,0 +1,398 @@
+	argument := Argument{
+		Name:     name,
+		HelpText: helpText,
+		Type:     DynamicListArgumentType,
+		Data:     &DynamicListArgument{FetchURL: url},
+	}
+	ad.Arguments = append(ad.Arguments, &argument)
+}
+
+// Equals method checks if command is the same.
+func (ad *AutocompleteData) Equals(command *AutocompleteData) bool {
+	if !(ad.Trigger == command.Trigger && ad.HelpText == command.HelpText && ad.RoleID == command.RoleID) {
+		return false
+	}
+	if len(ad.Arguments) != len(command.Arguments) || len(ad.SubCommands) != len(command.SubCommands) {
+		return false
+	}
+	for i := range ad.Arguments {
+		if ad.Arguments[i].Name != command.Arguments[i].Name ||
+			ad.Arguments[i].HelpText != command.Arguments[i].HelpText ||
+			ad.Arguments[i].Type != command.Arguments[i].Type ||
+			!reflect.DeepEqual(ad.Arguments[i].Data, command.Arguments[i].Data) {
+			return false
+		}
+	}
+	for i := range ad.SubCommands {
+		if !ad.SubCommands[i].Equals(command.SubCommands[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// IsValid method checks if autocomplete data is valid.
+func (ad *AutocompleteData) IsValid() error {
+	if ad.Trigger == "" {
+		return errors.New("An empty command name in the autocomplete data")
+	}
+	roles := []string{SYSTEM_ADMIN_ROLE_ID, SYSTEM_USER_ROLE_ID}
+	if stringNotInSlice(ad.RoleID, roles) {
+		return errors.New("Wrong role in the autocomplete data")
+	}
+	if len(ad.Arguments) > 0 {
+		namedArgumentIndex := -1
+		for i, arg := range ad.Arguments {
+			if !(arg.Name == "") { // it's a named argument
+				if namedArgumentIndex == -1 { // first named argument
+					namedArgumentIndex = i
+				}
+			} else { // it's a positional argument
+				if namedArgumentIndex != -1 {
+					return errors.New("Named argument should not be before positional argument")
+				}
+			}
+			if arg.Type == DynamicListArgumentType {
+				DynamicList, ok := arg.Data.(*DynamicListArgument)
+				if !ok {
+					return errors.New("Not a proper DynamicList type argument")
+				}
+				_, err := url.ParseRequestURI(DynamicList.FetchURL)
+				if err != nil {
+					return errors.Wrapf(err, "FetchURL is not a proper url")
+				}
+			} else if arg.Type == StaticListArgumentType {
+				StaticList, ok := arg.Data.(*StaticListArgument)
+				if !ok {
+					return errors.New("Not a proper StaticList type argument")
+				}
+				for _, arg := range StaticList.PossibleArguments {
+					if arg.Item == "" {
+						return errors.New("Possible argument name not set in StaticList argument")
+					}
+				}
+			} else if arg.Type == TextInputArgumentType {
+				if _, ok := arg.Data.(*TextInputArgument); !ok {
+					return errors.New("Not a proper TextInput type argument")
+				}
+			}
+		}
+	}
+	for _, command := range ad.SubCommands {
+		err := command.IsValid()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ToJSON encodes AutocompleteData struct to the json
+func (ad *AutocompleteData) ToJSON() ([]byte, error) {
```

Comment:
A little confused, why not `json.[Un]marshaler`? Maybe a convention in `model` I know nothing about?

Reply:
Looks like a convention, most `model` structs use `ToJSON`, `...FromJSON` wrappers.